src/ore/src/id_gen.rs

```diff
@@ -94,6 +107,14 @@ where
     pub fn free(&self, id: T) {
         let mut inner = self.0.lock().expect("lock poisoned");
         inner.free.push_back(id);
+        let _ = inner.allocated.remove(&id);
+    }
+
+    /// Marks an ID as allocated and prevents `IdAllocator` from allocating it
+    /// in the future.
+    pub fn mark_allocated(&self, id: T) {
```

Comment:
I think this needs to return a `bool` that indicates whether the ID allocator has already allocated `id`—i.e., if `self.next` is greater than `id` and `id` is not in `self.free`.

Reply:
I'm going to think about this overnight, but I'm not sure I understand why this would be needed.

Reply:
Ok I see what you're saying. I not sure if `self.next > id && `self.free.contains(id)` would be enough, because it's possible that `id` is outside the original range of allocator. In that case the above would be true but it would not have been previously allocated.

Though we may still want to return false because it's outside of the original range.